marimo/_ast/visitor.py

```diff
@@ -29,6 +29,10 @@ def __post_init__(self) -> None:
 class VariableData:
     kind: Literal["function", "class", "import", "variable"] = "variable"
 
+    # If kind == function or class, it may be dependent on externally defined
+    # variables.
+    required_refs: set[Name] = field(default_factory=set)
```

Comment:
Oh, this is interesting -- I was just thinking how it would be useful to have this information for functions and classes. Is it correct to say that if `required_refs` is the empty set, then the function or class is pure?

Reply:
It's not guaranteed to be pure since you might pass in a file descriptor or do something like `__import__("requests")`, but it's a good indication that the function _might_ be pure.